solve.ts

```diff
@@ -0,0 +1,198 @@
+export type voltage = {
+  type: "voltage";
+  V: number;
+}
+
+export type resistor = {
+  type: "resistor";
+  R: number;
+}
+
+export type diode = {
+  type: "diode";
+  Vd: number;
+}
+
+export type component = voltage | resistor | diode;
+export type node = number;
+
+export type placed = {
+  c: component;
+  p: node;
+  q: node;
+}
+
+export type solution = {
+  voltages: number[];
+  currents: number[];
+}
+
+// A should be a matrix with N rows (first index) and N columns (second index)
+// A is modified in place to contain both L & U
+// the permutation array is returned
+export function factor_plu(A: number[][]): number[] {
+  const N = A.length;
+
+  const permutations: number[] = [];
+  for (let i = 0; i < N; i++) {
+    permutations.push(i);
+  }
+
+  for (let k = 0; k < N; k++) {
+    // pick pivot for row k
+    let p = 0;
+    let k_swap = k;
+    for (let i = k; i < N; i++) {
+      if (Math.abs(A[i][k]) > p) {
+        p = Math.abs(A[i][k]);
+        k_swap = i;
+      }
+    }
+    if (p == 0) {
+      throw "singular matrix";
+    }
+    p = permutations[k];
+    permutations[k] = permutations[k_swap];
+    permutations[k_swap] = p;
+
+    // elimination
+    for (let i = 0; i < N; i++) {
+      let a = A[k][i];
```

Comment:
- `const a`
- comment "elimination" should go on the next block, not this one, which swaps rows